cukinia

```diff
@@ -891,6 +897,66 @@ _cukinia_knoerror()
 	return $ret
 }
 
+# _cukinia_network_interface_up: check if a network interface is up
+# arg1: interface name
+_cukinia_network_interface_up() {
+	local interface="$1"
+
+	_cukinia_prepare "Checking if interface \"$interface\" is ${__not:+NOT }up"
+
+	ip link show "$interface" up | grep -q 'state UP' >/dev/null 2>&1 && return 0 || return 1
+}
+
+# _cukinia_network_interface_ipv4: check if an interface has an IPv4 address
+# arg1: interface name
+_cukinia_network_interface_ipv4() {
+	local interface="$1"
+
+	_cukinia_prepare "Checking if interface \"$interface\" has ${__not:+NO }IPv4 address"
+
+	ip -4 addr show dev "$interface" | grep -q inet && return 0 || return 1
+}
+
+# _cukinia_network_interface_global_ipv6: check if a network interface has a global IPv6 address
+# arg1: interface name
+_cukinia_network_interface_global_ipv6() {
+	local interface="$1"
+
+	_cukinia_prepare "Checking if interface \"$interface\" has ${__not:+NO }global IPv6 address"
+
+	ip -o -6 address show dev "$interface" scope global | grep -q "inet6" && return 0 || return 1
+}
+
+# _cukinia_network_interface_dhcp_ipv4: check if an interface has an IPv4 address assigned via DHCP
+# arg1: interface name
+_cukinia_network_interface_dhcp_ipv4() {
+	local interface="$1"
+
+	_cukinia_prepare "Checking if interface \"$interface\" has ${__not:+NO }IPv4 address assigned via DHCP"
+
+	ip -4 addr show dev "$interface" | grep -q 'dynamic' && return 0 || return 1
```

Comment:
I think we're trying to be too smart by assuming that "dynamic" implies DHCP. AFAIK, this flag only means that the interface should lose its IP address if it is turned down. It is possible that some DHCP implementation don't set this bit (it is the case on my desktop machine) or that non-DHCP configuration tool set this bit. I think it is better to play it safe here as we're adding built-in commands.

We should either rename this cukinia command to `network_interface_dynamic_ipv4` or better, make a generic `network_interface_ipv4_flag` that greps for a flag passed as a parameter.

Reply:
No There is no assumption. DHCP client sets the dynamic flag and based on that this logic is implemented. IMO whenever interface gets IP from DHCP then dynamic flag should be set other wise "ip addr" command will not show correct information. (Desktop setup)If interface has IP address from DHCP but not dynamic(static), Not sure if this is the correct information or usecase. Which dhcp client you are running?

Reply:
I'm using NetworkManager.

I believe that "dynamic" means "the IFA_F_PERMANENT flag was not set for this address" [reference](https://github.com/shemminger/iproute2/blob/main/ip/ipaddress.c#L1460). The only information I'm able to find on IFA_F_PERMANENT are from Linux's [route netlink man page](https://man7.org/linux/man-pages/man7/rtnetlink.7.html) and [this RFC](https://datatracker.ietf.org/doc/html/rfc3549). None of those two sources mention DHCP but rather a "permanent address" and a "stateless autoconfiguration".

Reply:
OK. I think we can rename the function to "network_interface_dynamic_ipv4".

Reply:
That would make me happy =)